// iobeam-client/es5/resources/Device.js
"use strict";

var Utils = require("../utils/Utils");

/**
 * Represents a device.
 * @constructor
 * @param {string} id - Device id
 * @param {string} [name] - Human readable device names
 * @param {type} [type] - Device type
 * @returns {Device} A Device object with the given parameters.
 */
function Device(id, name, type, created) {
    var _id = id || null;
    if (_id) {
        Utils.assertValidDeviceId(_id);
    }
    var _name = name || null;
    var _type = type || null;
    var _created = created || null;

    /**
     * Get the device's created date.
     * @returns {string} Creation date as a string, null if not set.
     */
    this.getCreated = function () {
        return _created;
    };

    /**
     * Get the device's id.
     * @returns {string} Device id.
     */
    this.getId = function () {
        return _id;
    };

    /**
     * Get the device's name.
     * @returns {string} Device name, null if not set.
     */
    this.getName = function () {
        return _name;
    };

    /**
     * Get the device's type.
     * @returns {string} Device type, null if not set.
     */
    this.getType = function () {
        return _type;
    };

    this.toJSON = function () {
        var ret = { id: _id };
        if (_name) {
            ret.name = _name;
        }
        if (_type) {
            ret.type = _type;
        }
        if (_created) {
            ret.created = _created;
        }
        return ret;
    };
}

module.exports = Device;
