// iobeam-client/es5/endpoints/Exports.js
"use strict";

var RequestResults = require("../constants/RequestResults");
var Utils = require("../utils/Utils");

var _token = null;
var _requester = null;

function setOpt(request, opts, name) {
    var temp = {};
    if (opts[name]) {
        temp[name] = opts[name];
        request.query(temp);
    }
}

module.exports = {

    initialize: function initialize(token, requester) {
        Utils.assertValidToken(token);
        Utils.assertValidRequester(requester);
        _token = token;
        _requester = requester;
    },

    query: function query(projectId, deviceId, seriesName, callback, options) {
        Utils.assertValidToken(_token);
        Utils.assertValidProjectId(projectId);
        var did = deviceId || "all";
        var sname = seriesName || "all";
        var opts = options || {};
        if (did !== "all") {
            Utils.assertValidDeviceId(did);
        }
        var endpoint = "/exports/" + projectId + "/" + did + "/" + sname;
        var URL = _requester.getFullEndpoint(endpoint);
        var context = {
            projectId: projectId,
            deviceId: deviceId,
            seriesName: seriesName,
            options: opts
        };

        var req = _requester.getRequest(URL, _token);
        setOpt(req, opts, "limit");
        setOpt(req, opts, "from");
        setOpt(req, opts, "to");
        setOpt(req, opts, "output");
        setOpt(req, opts, "group_by");
        setOpt(req, opts, "operator");
        setOpt(req, opts, "timefmt");

        var bodyHandler = function bodyHandler(resp, body, status) {
            if (status === RequestResults.SUCCESS) {
                resp.body = body;
            } else if (body && body.errors) {
                resp.error = body.errors[0].message;
            }
        };
        var innerCb = Utils.createInnerCb(callback, context, bodyHandler);
        _requester.execute(req, innerCb);
    }
};
