// iobeam-client/es5/endpoints/Imports.js
"use strict";

var RequestResults = require("../constants/RequestResults");
var Utils = require("../utils/Utils");

var _token = null;
var _requester = null;

module.exports = {

    /**
     * Initialize the Imports module.
     * @param {string} token - Project token to use when making API calls
     * @param {Requester} requester - Object that performs the API calls via HTTP
     */
    initialize: function initialize(token, requester) {
        Utils.assertValidToken(token);
        Utils.assertValidRequester(requester);
        _token = token;
        _requester = requester;
    },

    /**
     * Perform API call to send data in series format
     * @param {int} projectId - Project ID this data belongs to
     * @param {string} deviceId - Device ID this data belongs to
     * @param {object} dataSet - Data organized by series names as properties and
     * arrays of data points.
     * @param {function} callback - Function to be called after request returns.
     */
    import: function _import(projectId, deviceId, dataSet, callback) {
        Utils.assertValidToken(_token);
        Utils.assertValidProjectId(projectId);
        Utils.assertValidDeviceId(deviceId);
        var URL = _requester.getFullEndpoint("/imports");

        var reqBody = {
            project_id: parseInt(projectId),
            device_id: deviceId,
            sources: []
        };

        for (var series in dataSet) {
            if (dataSet.hasOwnProperty(series)) {
                (function () {
                    var seriesTemp = { name: series, data: [] };
                    dataSet[series].forEach(function (dp) {
                        var temp = { time: dp.timestamp, value: dp.value };
                        seriesTemp.data.push(temp);
                    });
                    reqBody.sources.push(seriesTemp);
                })();
            }
        }

        var req = _requester.postRequest(URL, reqBody, _token);
        var innerCb = Utils.createInnerCb(callback, null, function () {});
        _requester.execute(req, innerCb);
    },

    /**
     * Performs API call to send data in store/table format.
     * @param {int} projectId - Project ID this data belongs to
     * @param {string} deviceId - Device ID this data belongs to
     * @param {DataStore} store - Data in table format
     * @param {function} callback - Function to be called after request returns.
     */
    importBatch: function importBatch(projectId, deviceId, store, callback) {
        Utils.assertValidToken(_token);
        Utils.assertValidProjectId(projectId);
        Utils.assertValidDeviceId(deviceId);
        if (store.rows().length === 0) {
            if (Utils.isCallback(callback)) {
                var resp = Utils.getDefaultApiResp(RequestResults.SUCCESS, { status: 200 });
                callback(resp);
            }
            return;
        }

        var context = {
            projectId: projectId,
            deviceId: deviceId,
            store: store.snapshot()
        };
        var URL = _requester.getFullEndpoint("/imports?fmt=table");
        var reqBody = {
            project_id: parseInt(projectId),
            device_id: deviceId
        };

        var fields = ["time"];
        store.fields().forEach(function (f) {
            fields.push(f);
        });

        var rows = store.rows();
        var data = [];
        rows.forEach(function (r) {
            var temp = [];
            fields.forEach(function (f) {
                temp.push(r[f]);
            });
            data.push(temp);
        });

        reqBody.sources = {
            fields: fields,
            data: data
        };

        var req = _requester.postRequest(URL, reqBody, _token);
        var bodyHandler = function bodyHandler(resp, body, status) {
            if (status !== RequestResults.SUCCESS) {
                if (body && body.errors) {
                    resp.error = body.errors[0];
                }
            }
        };
        var innerCb = Utils.createInnerCb(callback, context, bodyHandler);
        _requester.execute(req, innerCb);
    }
};
